storage/sql_test.go

```diff
@@ -41,3 +44,473 @@ func TestSchemaIsCurrent(t *testing.T) {
 		})
 	}
 }
+
+func TestLeaseStateChanges(t *testing.T) {
+	if testing.Short() || !testutil.DatabaseAvailable() {
+		t.Skip("short testing requested or VISOR_TEST_DB not set")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	db, cleanup, err := testutil.WaitForExclusiveDatabase(ctx, t)
+	require.NoError(t, err)
+	defer cleanup()
+
+	truncateVisorProcessingTables(t, db)
+
+	indexedTipsets := visor.ProcessingStateChangeList{
+		// Unclaimed, incomplete tipset
+		{
+			TipSet:  "cid0a,cid0b",
+			Height:  0,
+			AddedAt: testutil.KnownTime,
+		},
+
+		// Unclaimed, incomplete tipset
+		{
+			TipSet:  "cid1a",
+			Height:  1,
+			AddedAt: testutil.KnownTime,
+		},
+
+		// Unclaimed, incomplete tipset
+		{
+			TipSet:  "cid2a,cid2b,cid2c",
+			Height:  2,
+			AddedAt: testutil.KnownTime,
+		},
+
+		// Unclaimed, incomplete tipset
+		{
+			TipSet:  "cid3a",
+			Height:  3,
+			AddedAt: testutil.KnownTime,
+		},
+
+		// TipSet completed with stale claim
+		{
+			TipSet:       "cid4",
+			Height:       4,
+			AddedAt:      testutil.KnownTime,
+			ClaimedUntil: testutil.KnownTime.Add(-time.Minute * 15),
+			CompletedAt:  testutil.KnownTime.Add(-time.Minute * 5),
+		},
+
+		// TipSet claimed by another process that has expired
+		{
+			TipSet:       "cid5",
+			Height:       5,
+			AddedAt:      testutil.KnownTime,
+			ClaimedUntil: testutil.KnownTime.Add(-time.Minute * 5),
+		},
+
+		// TipSet claimed by another process
+		{
+			TipSet:       "cid6a,cid6b",
+			Height:       6,
+			AddedAt:      testutil.KnownTime,
+			ClaimedUntil: testutil.KnownTime.Add(time.Minute * 15),
+		},
+	}
+
+	if err := db.RunInTransaction(ctx, func(tx *pg.Tx) error {
+		return indexedTipsets.PersistWithTx(ctx, tx)
+	}); err != nil {
+		t.Fatalf("persisting indexed blocks: %v", err)
+	}
+
+	const batchSize = 3
+
+	claimUntil := testutil.KnownTime.Add(time.Minute * 10)
+	d := &Database{DB: db}
+
+	claimed, err := d.LeaseStateChanges(ctx, claimUntil, batchSize, 500)
+	require.NoError(t, err)
+	require.Equal(t, batchSize, len(claimed), "number of claimed blocks")
+
+	// TipSets are selected in descending height order, ignoring completed and claimed tipset
+	assert.Equal(t, "cid5", claimed[0].TipSet, "first claimed tipset")
+	assert.Equal(t, "cid3a", claimed[1].TipSet, "second claimed tipset")
+	assert.Equal(t, "cid2a,cid2b,cid2c", claimed[2].TipSet, "third claimed tipset")
+
+	// Check the database contains the leases
+	var count int
+	_, err = db.QueryOne(pg.Scan(&count), `SELECT COUNT(*) FROM visor_processing_statechanges WHERE claimed_until=?`, claimUntil)
+	require.NoError(t, err)
+	assert.Equal(t, batchSize, count)
+}
```

Comment:
We discussed this in chat. This will create extra work and errors for visor. If we can be sure that errors don't impact other processing results from persisting to the DB, then it's probably okay.